src/main_loop.rs

```diff
@@ -228,137 +231,156 @@ impl SyncState {
+    /// The decision whether a potential peer is identical to the current node
+    /// is made _only_ based on the [`InstanceId`]. This means that different
+    /// nodes that run on the same computer _can_ connect to each other.
+    ///
+    ///   * acquires `global_state_lock` for read
+    //
+    // A note on the design: The method first declares various relevant filters
+    // as well as a way to compare two peer candidates. The filters and the
+    // comparison function are then used in a simple iterator-combinator chain
+    // to select the candidate.
+    //
+    // The design's intention is to simplify future changes: declaring a new
+    // filter as well as modifying or dropping existing filters has little to no
+    // impact on other filters. Similarly, modifying the comparison function
+    // does not require changes to the filters.
+    async fn peer_candidate(&self) -> Option<PeerCandidate> {
+        // trade-off: hold the lock for a while, don't clone a bunch of stuff.
+        let networking_state = &self.global_state_lock.lock_guard().await.net;
+
+        let is_self = |candidate: &PeerCandidate| candidate.id == networking_state.instance_id;
+        let is_peer = |candidate: &PeerCandidate| {
+            networking_state
+                .peer_map
+                .values()
+                .any(|peer| candidate.id == peer.instance_id())
         };
+        let is_bootstrap_node = |candidate: &PeerCandidate| {
+            let candidate_bootstrap_status = networking_state
+                .bootstrap_status
+                .get(&candidate.address)
+                .map(|bootstrap_info| bootstrap_info.status)
+                .unwrap_or_default();
+            candidate_bootstrap_status == BootstrapStatus::Bootstrap
+        };
+        let self_is_well_connected = self.is_well_connected().await;
 
-        // Get the candidate list with the highest distance
-        let max_distance_candidates = candidates.iter().max_by_key(|pp| pp.1.distance);
+        self.candidates
+            .values()
+            .filter(|candidate| !is_self(candidate))
+            .filter(|candidate| !is_peer(candidate))
+            .filter(|candidate| !self_is_well_connected || !is_bootstrap_node(candidate))
+            .max_by(|l, r| Self::candidate_ordering(networking_state.peer_map.values(), l, r))
+            .copied()
+    }
 
-        // Pick a random candidate from the appropriate candidates
-        let mut rng = rand::rng();
-        max_distance_candidates
-            .iter()
-            .choose(&mut rng)
-            .map(|x| (x.0.to_owned(), x.1.distance))
+    ///   * acquires `global_state_lock` for read
+    async fn is_well_connected(&self) -> bool {
+        /// A node with more peers than this ratio of its maximum number of
+        /// peers is considered well-connected.
+        const WELL_CONNECTED_RATIO: f32 = 0.8;
+
+        let curr_num_peers = self.global_state_lock.lock_guard().await.net.peer_map.len();
```

Comment:
This is the 2nd time a read lock is being grabbed by  this task. This will lead to a deadlock if any task requests a write lock between the 1st lock being given and the 2nd being given. We *really* should have a way to test that the same task is never requested twice.

This complicated pitfall is why I prefer to get all relevant data from global state upfront, as far out in the callstack as possible. It does lead to more complicated function interfaces, but on the upside, you functions can become purely functional and locking issues are much easier to reason about.